testdata/p4_16_samples/array3.p4

```diff
@@ -0,0 +1,30 @@
+control proto<P>(inout P pkt);
+package top<P>(proto<P> p);
+
+struct S {
+    bit<16> f1;
+    bit<8>[16] nested_arr;
+    bit<16> f2;
+}
+
+header data_t {
+    S[16]  arr;
+    bit<4>     w;
+    bit<4>     x;
+    bit<4>     y;
+    bit<4>     z;
+}
```

Comment:
Should we also test the other case I mentioned (i.e.
```
header data_t {
    bit<32>     f1;
    bit<32>     f2;
    bit<16>     h1;
    bit<16>     h2;
    bit<8>[16][16]  arr;
}
...
    hdr.data.arr[x][y] = ...;
    ... = hdr.data.arr[z][w];
...
```
) (assuming that the grammar allows it)?

Reply:
This code blows up too much.  Even with only 2 indexes in a single statement, it gets 256 variants which expands the memory a lot and takes 5-10 minutes to run the test.  Trying to do 3+ really blows up.

There are probably things that could be improved in HSIndexSimplifier pass to not blow up quite as badly, but it is tricky and probably beyond the scope of this change.

Reply:
Is there a reason we do not want to explicitly disallow this case (by detecting it and emitting an error message)? Is it because not all backends run `HSIndexSimplifier`? If that's the case, then why not detect it inside of `HSIndexSimplifier` pass, before `HSIndexContretizer` (and exit with an error instead of running `HSIndexContretizer` if it is detected)?

Reply:
The same problem can happen even without 2D arrays if they are large enough.  The best "quick-fix" for HSIndexSimplifier is probably to put in a limit to the number of branch leaves inserted at 100 or so -- that would avoid any case that had a huge blowup independent of dimensions and such,

Reply:
Ugh, `HSIndexSimplifier` is just broken in how it deals with multiple indexes on arrays/header stacks.  It basically goes over things multiple times, hoping everything gets expanded, which may or may not happen (and may create many unneeded duplicates of things).

Reply:
I've add a limit to the expansion in HSIndexSimplifier -- it now tracks how many additional things it is adding and errors out at an arch-specified limit (defaults to 1000).

I don't like the code of this pass -- it is pretty convoluted and probably broken for some corner cases we don't check.  Should be replaced by a simpler pass that does not create and apply sub-passes dynamically and does a better job.